### PR TITLE
Physical skeleton improvement and bug fix

### LINKED ISSUE #19002: Ragdoll / Physical Bones issue
As described in #11973
I've attached an example project below:
[Ragdoll Test.zip](https://github.com/godotengine/godot/files/2016597/Ragdoll.Test.zip)

Example contains 1 scene with 2 skeletons.
First example - a simple worm skeleton that contorts wildly.
Second Example - Robot from 3D platformer, by default I've set the physics simulation to just his arms. They don't stop moving, as if he's very excited about something.
If you hit Enter (ui_accept) the rest of his bones are activated, and he crumples to the floor in quite an unnatural way, not like in the example posted by Andrea (https://godotengine.org/article/godot-ragdoll-system)

### LINKED ISSUE #22823: Bullet warning when PhysicalBone is deleted
**Godot version:**

3.1 master -> f12cb82

**OS/device including version:**

Windows 10 Home

**Issue description:**

When a PhysicalBone is deleted, a warning appears:

```
WARNING: RigidBodyBullet::assert_no_constraints: A body with a joints is destroyed. Please check the implementation in order to destroy the joint before the body.
At: modules\bullet\rigid_body_bullet.cpp:442   
```

**Steps to reproduce:**
`queue_free` a `PhysicalBone` node (or any of its parents) that is attached to a skeleton.

**Minimal reproduction project:**

[physical-bones-crash.zip](https://github.com/godotengine/godot/files/2430780/physical-bones-crash.zip)

Same project as #22538 - the warnings happen, but currently the editor crashes afterwards for me.

### LINKED ISSUE #24968: Bullet error spam after changing scenes when the old scene contains an entity with physical bones
**Godot version:**
Godot 3.1 Alpha 5

**OS/device including version:**
Kubunu 18.10

**Issue description:**
I stumbled upon this bug when implementing a level loader and messing around with the ragdoll features. Godot crashes after changing the scene when the old scene contains an entity with physical bones.

I have attached a minimal reproduction project containing 2 scenes. The main `Game` scene has a `Timer` node and a script attached to it that changes the current scene after 3 seconds.

The scene changing logic is very simple. It basically removes the old scene, calls `queue_free` and adds an instance of the new scene. 

```
func _change_scene():
	# remove the old scene
	var old_scene = get_node("Scene")
	remove_child(old_scene)
	old_scene.queue_free()
	
	# load and add new scene
	var scene = load("res://Scene_2.tscn").instance()
	scene.name = "Scene"
	add_child(scene)
```

The console output is also littered with the following warning coming  from bulletphysics. It seems related https://github.com/godotengine/godot/issues/22823:
```
modules/bullet/rigid_body_bullet.cpp:463 - A body with a joints is destroyed. Please check the implementation in order to destroy the joint before the body.
```

**NB:**
* On the rare occasion it doesn't crash,  run it a few more times.
* Funnily enough, removing the entity with the physical bones on its own and calling queue_free doesn't crash Godot, only when changing the scenes.

**Steps to reproduce:**
Run the attached minimal reproduction project.

**Minimal reproduction project:**
[Godot_Change_Scene.zip](https://github.com/godotengine/godot/files/2753095/Godot_Change_Scene.zip)

### LINKED ISSUE #25564: Let PhysicalBone detect collisions, and allow application of impulses
**Godot version:**
`master` / f32c826924

**OS/device including version:**
Manjaro Linux 18.0.2

**Issue description:** _(This is a feature request)_
What I'm trying to achieve is to make a 3d character react to a hit impact in a more realistical manner. To be specific, imagine a sword fighting game in which you can attack your enemy and hit his various body parts with your sword.

To be able to achieve such an effect, you'd probably need to know exactly where (i.e. global 3d position in `Vector3`) the sword hit, and strength and direction of the impact (i.e. a force represented as `Vector3`).

Once you get that information somehow, you need to apply that force to the appropriate part of the enemy character's body. I imagine such a character would have a ragdoll setup, consisting of `PhysicalBone` nodes in a 'sleeping' status then need to get 'awakened' (i.e. by invoking `Skeleton.PhysicalBonesStartSimulation`) before applying the aforementioned force data (i.e. like `RigidBody.IntegrateForce`).

However, if I'm not mistaken, neither of those is possible with the current status of Godot. First off, there is no signal to detect collision in `PhysicalBone` or `CollisionShape` so we can't take the more natural approach of letting the object that is being hit to decide what could be a proper reaction to the impact.

And even if we can make the sword a `RigidBody`, driving it by applying forces would be awfully impractical, since the movement of the weapon is usually determined by what animation the character plays when it attacks, rather than based on some complex procedural, physics calculations.

In Unity, in comparison, there is a `OnCollisionEnter(Collision)` callback which I believe to be usable from any collider component. I think it'd be much better if Godot has an equivalent signal in either `CollisionShape` or `PhysicsObject` node, so that we can gather such information like exactly where the object was hit or how hard and more.

Also, it doesn't seem to be possible to apply force to exact location on the target rig with proper intensity and direction. We can make the character `RigidBody` and use `IntegrateForce`, of course, but it'd probably make the whole body flying when I hit its arm, for instance, instead of just making the arm thrown back as it is expected when it gets hit by a solid object.

Ideally, it'd be better if either we can make `PhysicalBone` to inherit from `RigidBody`, or pull up methods like `IntegrateForce` to their common ancestor, so we can interact with the former in the above mentioned scenario.

### LINKED ISSUE #26618: Move joint does not work
**Godot version:**
Godot 3.1 Beta 10

**OS/device including version:**
GTX970, Windows 10 64bit, Nvidia driver 419.17

**Issue description:**
The "Move joint" button in the editor does not work. I am trying to do the ragdoll tutorial from here 
https://docs.godotengine.org/en/latest/tutorials/physics/ragdoll_system.html . When I tried to do the joint adjustment for the "physical bone head" even though I clicked on the "Move joint" button, it will actually rotate / move the physical bone and not the joint like in the tutorial.

**Steps to reproduce:**
1)Download Godot 3.1 Beta 10, 
2)Open the 3d platformer template
3)Open the player scene
4)Create physical bones and follow the tutorial from here https://docs.godotengine.org/en/latest/tutorials/physics/ragdoll_system.html 
5)When trying to do the joint adjustment , the "move joint" feature or button does not work. It will only rotate or move the physical bones and not the joint.

**Minimal reproduction project:**
Download 3d platformer example from the asset library

### LINKED ISSUE #28707: ragdoll simulation twitches in simple humanoid ragdoll
**Godot version:**
v3.2.dev.custom_build.2931b4d

**OS/device including version:**
OS: Ubuntu 18.04.2

**Issue description:**
ragdoll simulation twitches in simple humanoid ragdoll with conejoint as constraint (except for the head with hingejoint), and default settings for the physical bones and the joint constraint (except swing span)  [video with the issue](https://streamable.com/yh958)

**Minimal reproduction project:**
[testr.zip](https://github.com/godotengine/godot/files/3145742/testr.zip)
(Run simulation from playground scene. Ignore actor_def scene)

### PR DESCRIPTION
**Pull request by AndreaCatania**
This PR is a rework of the physical bone that aim to fix all the current open issues and also improve the current Physical Bone by adding stability and control.

The current Physical Bone implementation is using the normal rigid bodies and joints, and it works good but it become unstable if the complexity of the skeleton is not trivial, for this reason the Featherstone algorithm got implemented thus allowing to have not only stable body but also full control of it.

When I talk about control I mean that with this PR is not only possible to create ragdoll of not animated body but is also possible control its movement by adding forces like the following:

ezgif com-video-to-gif

This dummy is not animated using the animator, but it's animated controlling the motor of each bone, and the responsible code is simply:
```
func _process(delta):

	if Input.is_action_just_pressed("push"):

		left_arm.motor_set_rotation_target(Vector3(0, 0, 90))
		right_arm.motor_set_rotation_target(Vector3(0, 0, 90))
		left_forearm.motor_set_position_target(0)
		right_forearm.motor_set_position_target(0)

	elif Input.is_action_just_released("push"):

		left_arm.motor_set_rotation_target(Vector3(0, 0, 0))
		right_arm.motor_set_rotation_target(Vector3(0, 0, 0))
		left_forearm.motor_set_position_target(120)
		right_forearm.motor_set_position_target(120)
```

You don't have to deal with complex force system, but instead you can control it just by changing motor target and or velocity and some other parameters from the editor.

Thanks to this new PR is now possible to create, the a lot discussed, active ragdoll directly in GDscript, or is possible to create a completely physical walking robot, or simply is possible to have a stable ragdoll with dumpers and springs.

But test it yourself to see the new features, and if you want you can use this project that I've to used to develop this new feature:
[Multibody.zip](https://github.com/godotengine/godot/files/3182509/Multibody.zip)

This PR Fix:
Fixes #28707
Fixes #26618
Fixes #25564
Fixes #24968
Fixes #22823
Fixes #19002

### CODE REVIEW

File: thirdparty/bullet/BulletDynamics/Featherstone/btMultiBody.cpp [hunk 1]
```diff
@@ -3,7 +3,7 @@
  *   Class representing an articulated rigid body. Stores the body's
  *   current state, allows forces and torques to be set, handles
  *   timestepping and implements Featherstone's algorithm.
- *   
+ *
```
Comment (akien-mga) [inline thirdparty/bullet/BulletDynamics/Featherstone/btMultiBody.cpp]: Your IDE removed trailing spaces in thirdparty code, could you remove those changes?

File: thirdparty/bullet/BulletDynamics/Featherstone/btMultiBody.cpp [hunk 2]
```diff
@@ -343,6 +343,7 @@ void btMultiBody::finalizeMultiDof()
 	m_deltaV.resize(6 + m_dofCount);
 	m_realBuf.resize(6 + m_dofCount + m_dofCount * m_dofCount + 6 + m_dofCount);  //m_dofCount for joint-space vels + m_dofCount^2 for "D" matrices + delta-pos vector (6 base "vels" + joint "vels")
 	m_vectorBuf.resize(2 * m_dofCount);                                           //two 3-vectors (i.e. one six-vector) for each system dof	("h" matrices)
+	m_matrixBuf.resize(m_links.size() + 1);
```
Comment (akien-mga) [inline thirdparty/bullet/BulletDynamics/Featherstone/btMultiBody.cpp]: Why is this change needed? It should also be PR'ed upstream if it's a bugfix.
Merging that also means that we're no longer compatible with Bullet 2.88 which is the current stable release, so if we can handle this on our end without modifying Bullet for now, that would be best.

Comment (AndreaCatania) [reply thirdparty/bullet/BulletDynamics/Featherstone/btMultiBody.cpp]: Nope is not possible to manage it internally, I did a PR on the Bullet side also: https://github.com/bulletphysics/bullet3/pull/2236 but in order to be able to use it ASAP I did it in this way

File: thirdparty/bullet/BulletDynamics/Featherstone/btMultiBody.cpp (hunk 1 above)
Comment (AndreaCatania) [reply thirdparty/bullet/BulletDynamics/Featherstone/btMultiBody.cpp]: Yes

**Comment by Zireael07** [comment]
@AndreaCatania: I saw you showcase ragdoll/skeleton with this dummy, where can we get the assets? Is it on the asset lib? Getting skeletons working with Blender -> Godot can be tricky, it would be great to have an easily available working sample...

**Comment by AndreaCatania** [comment]
@Zireael07 do you mean that you want to try it by yourself?
If yes, is the attached project not working?
If not, sorry but I didn't understood what you mean, can you please rephrase the question?

**Comment by Zireael07** [comment]
I mean I could never get the skeleton exported from Blender to work, so if your dummy asset were generally available on the asset lib, it would be easier for people to get started with skeletons in Godot.

It is just a tangential question since I saw you demo'ing various skeleton stuff with always the same dummy, the dummy is therefore guaranteed to work.

**Comment by AndreaCatania** [comment]
Oh ok, To be fair this is where you can download it: https://www.mixamo.com/ and for this precise work I've used the skeleton from this guy: #28707 but yes indeed would be useful to make the skeleton import more easy.

I'll try to look on that in the next days

**Comment by mysticfall** [comment]
I'm trying to test this branch but getting the following error when I run my project. Could you give me some help to resolve this problem?
```
E 0:00:02:0315   This multi body can't be removed from non multi body world
  <C Source>     modules/bullet/space_bullet.cpp:518 @ remove_armature()
```

**Comment by akien-mga** [comment]
#29418 is merged so you can rebase on master to fix CI.

**Comment by AndreaCatania** [comment]
Also this get merged https://github.com/bulletphysics/bullet3/pull/2279

This weekend I'll be able to work on this.

@mysticfall you want to active the multi body world in the settings. Check the physics tab in the project settings to active it.

**Comment by AndreaCatania** [comment]
Done!

File: editor/spatial_editor_gizmos.cpp [hunk 3]
```diff
@@ -1850,64 +1855,62 @@ void PhysicalBoneSpatialGizmoPlugin::redraw(EditorSpatialGizmo *p_gizmo) {
...
 					pb ? &points : NULL,
 					pbp ? &points : NULL);
+
 		} break;
-		case PhysicalBone::JOINT_TYPE_SLIDER: {
+		case PhysicalBone::JOINT_TYPE_SPHERICAL: {
 
-			const PhysicalBone::SliderJointData *sjd(static_cast<const PhysicalBone::SliderJointData *>(physical_bone->get_joint_data()));
-			JointSpatialGizmoPlugin::CreateSliderJointGizmo(
+			const PhysicalBone::SphericalJointData *sjd(static_cast<const PhysicalBone::SphericalJointData *>(physical_bone->get_joint_data()));
```
Comment (akien-mga) [inline editor/spatial_editor_gizmos.cpp]: This one is unused.

File: editor/spatial_editor_gizmos.cpp [hunk 4]
```diff
@@ -1850,64 +1855,62 @@ void PhysicalBoneSpatialGizmoPlugin::redraw(EditorSpatialGizmo *p_gizmo) {
...
-
-					sdofjd->axis_data[2].angular_limit_lower,
-					sdofjd->axis_data[2].angular_limit_upper,
-					sdofjd->axis_data[2].linear_limit_lower,
-					sdofjd->axis_data[2].linear_limit_upper,
-					sdofjd->axis_data[2].angular_limit_enabled,
-					sdofjd->axis_data[2].linear_limit_enabled,
+		case PhysicalBone::JOINT_TYPE_PLANAR: {
+
+			const PhysicalBone::PlanarJointData *pjd(static_cast<const PhysicalBone::PlanarJointData *>(physical_bone->get_joint_data()));
```
Comment (akien-mga) [inline editor/spatial_editor_gizmos.cpp]: Also unused.

**Comment by akien-mga** [comment]
Looks good to me overall. Couple unused variables which produce warnings, and `disable_3d` build is broken: https://travis-ci.org/godotengine/godot/jobs/545622746

Could you add some details on the changes in the body of the commit log? So that we have some information in `git log` on what changes are being done and why.

**Comment by akien-mga** [comment]
I force pushed an update to fix compilation warnings and build failure with `disable_3d=yes`.

I also synced the documentation so that we can see the actual API changes, which are compatibility breaking, so this will be difficult to merge for 3.2, will likely have to wait for 4.0.

I tried it quickly on the 3D platformer demo and the TPS demo, and they both throw a lot of error simply when opening their main scene:

- TPS demo:

```
ERROR: free: Invalid ID
   At: servers/physics/physics_server_sw.cpp:1414.
ERROR: shape_create: CylinderShape is not supported in GodotPhysics. Please switch to Bullet in the Project Settings.
   At: servers/physics/physics_server_sw.cpp:77.
ERROR: get: Condition ' !p_rid.is_valid() ' is true. returned: __null
   At: ./core/rid.h:153.
ERROR: shape_set_data: Condition ' !shape ' is true.
   At: servers/physics/physics_server_sw.cpp:107.
ERROR: free: Invalid ID
   At: servers/physics/physics_server_sw.cpp:1414.
ERROR: set_data: Condition ' !d.has("cell_size") ' is true.
   At: servers/physics/shape_sw.cpp:1650.
ERROR: free: Invalid ID
   At: servers/physics/physics_server_sw.cpp:1414.
ERROR: get: Condition ' !p_rid.is_valid() ' is true. returned: __null
   At: ./core/rid.h:153.
ERROR: body_attach_object_instance_id: Condition ' !body ' is true.
   At: servers/physics/physics_server_sw.cpp:679.
ERROR: free: Invalid ID
   At: servers/physics/physics_server_sw.cpp:1414.
ERROR: free: Invalid ID
   At: servers/physics/physics_server_sw.cpp:1414.
```

- 3D platformer:

```
ERROR: remove_armature: This multi body can't be removed from non multi body world
   At: modules/bullet/space_bullet.cpp:518.
ERROR: remove_armature: This multi body can't be removed from non multi body world
   At: modules/bullet/space_bullet.cpp:518.
ERROR: remove_armature: This multi body can't be removed from non multi body world
   At: modules/bullet/space_bullet.cpp:518.
ERROR: remove_armature: This multi body can't be removed from non multi body world
   At: modules/bullet/space_bullet.cpp:518.
ERROR: remove_armature: This multi body can't be removed from non multi body world
   At: modules/bullet/space_bullet.cpp:518.
WARNING: _notification: AnimationTreePlayer has been deprecated. Use AnimationTree instead.
   At: scene/animation/animation_tree_player.cpp:407.
ERROR: remove_armature: This multi body can't be removed from non multi body world
   At: modules/bullet/space_bullet.cpp:518.
ERROR: remove_armature: This multi body can't be removed from non multi body world
   At: modules/bullet/space_bullet.cpp:518.
ERROR: remove_armature: This multi body can't be removed from non multi body world
   At: modules/bullet/space_bullet.cpp:518.
ERROR: remove_armature: This multi body can't be removed from non multi body world
   At: modules/bullet/space_bullet.cpp:518.
ERROR: remove_armature: This multi body can't be removed from non multi body world
   At: modules/bullet/space_bullet.cpp:518.
```

**Comment by AndreaCatania** [comment]
The APIs doesn't changed too much but some things are not compatible since I've also improved a bit the APIs.

However the error that you got are mostly because you have to active the multibody world to use it.

**Comment by rodolforg** [comment]
Won't this PR be able to 3.2 (or 3.1.2)? :'(

**Comment by maystey** [comment]
@AndreaCatania I'm having trouble with the motor velocity targets. In all of my testing nothing seems to happen when I set a velocity target (in the inspector or with scripts). 

This includes the "Snake" actor in your demo project (I uncommented the line in the script that set the velocity). I increased the max impulse of the chosen physical bone, but still got nothing.

I don't see any errors for this, just no change in the simulation.

**Comment by GeorgeS2019** [comment]
@AndreaCatania @akien-mga Can we please re-visit this issue. 

Multibody physics is key to implementing more advanced IK not possible within the current IK implementation. 

Please advice how best to proceed. 
Is it possible to include this in the next 3.2.2 release?

Please also advice what to do when issue like this got lost during busy schedule when the team busy with GodotCon 2020.

**Comment by GeorgeS2019** [comment]
@AndreaCatania @akien-mga As Juan spoke recently, he see one of the focus points for godot 4.0 will be 3D. Is there a discussion platform where we could discuss what should be the godot 4.0 focus in terms of 3D features. From the feedback I gathered, I think IK, moving bones in editor (Tool mode), soft physics (perhaps using Flex, which again the issue of Cuda, can we have a way to turn on for 3D in desktop versus mobile?, the future android running on snapdragon will have superior AI engine, tapping into the soon-to-be industry standard mobile AI hardware support is also something needs to be discussed to strategic way attracting talented AI mobile developers into godot community).  For people who like to continue this discussion, I can be reached through the official discord [GeorgeFFM]

**Comment by vickylance** [comment]
@AndreaCatania  Is this fix by any chance going to be merged for 3.2.2 release?

**Comment by Calinou** [comment]
@vickylance There are many merge conflicts in this pull request, and 3.2.2 is very close to being released, so I doubt it.

**Comment by akien-mga** [comment]
@AndreaCatania What's the status on this PR for `master` (and possibly a version for `3.2`)?

**Comment by aaronfranke** [comment]
This PR has not received any new commits from the author for over a year and has many conflicts, closing.

If this is rebased and updated to work with the latest master branch, this can be re-opened or a new PR can be opened instead.

**Comment by AndreaCatania** [comment]
I'll rebase this work on master. Not sure if it's good to back port the code at this point to 3.2. I'll open another Pr.
